agent/engine/task_manager.go

```diff
@@ -732,8 +805,80 @@ func (mtask *managedTask) startContainerTransitions(transitionFunc containerTran
 	return anyCanTransition, transitions, reasons
 }
 
+// startResourceTransitions steps through each resource in the task and calls
+// the passed transition function when a transition should occur
+func (mtask *managedTask) startResourceTransitions(transitionFunc resourceTransitionFunc) (bool, map[string]string) {
+	anyCanTransition := false
+	transitions := make(map[string]string)
+	for _, res := range mtask.Resources {
+		knownStatus := res.GetKnownStatus()
+		desiredStatus := res.GetDesiredStatus()
+		if knownStatus >= desiredStatus {
+			seelog.Debugf("Managed task [%s]: resource [%s] has already transitioned to or beyond the desired status %s: %s",
+				mtask.Arn, res.GetName(), res.StatusString(desiredStatus), res.StatusString(knownStatus))
+			continue
+		}
+		anyCanTransition = true
+		transition := mtask.resourceNextState(res)
+		// If the resource is already in a transition, skip
+		if transition.actionRequired && !res.SetAppliedStatus(transition.nextState) {
+			// At least one resource is able to be moved forwards, so we're not deadlocked
+			continue
+		}
+		if !transition.actionRequired {
+			// no action is required for the transition, just set the known status without
+			// calling any transition function
+			go mtask.handleResourceChange(res, transition.nextState, nil)
+			continue
+		}
+		// At least one resource is able to be move forwards, so we're not deadlocked
+		transitions[res.GetName()] = transition.statusString
+		go transitionFunc(res, transition.nextState)
+	}
+
+	return anyCanTransition, transitions
+}
+
+// transitionResource calls applyResourceState, and then notifies the managed
+// task of the change. transitionResource is called by progressTask
+func (mtask *managedTask) transitionResource(resource taskresource.TaskResource,
+	to taskresource.ResourceStatus) {
+	engine := mtask.engine
+	err := mtask.applyResourceState(resource, to)
+
+	engine.processTasks.RLock()
+	_, ok := engine.managedTasks[mtask.Arn]
+	engine.processTasks.RUnlock()
+	if ok {
+		mtask.handleResourceChange(resource, to, err)
+	}
+}
+
+// applyResourceState moves the resource to the given state by calling the
+// function defined in the transitionFunctionMap for the state
+func (mtask *managedTask) applyResourceState(resource taskresource.TaskResource,
+	nextState taskresource.ResourceStatus) error {
+	ok, err := resource.ApplyTransition(nextState)
+	if !ok {
+		seelog.Criticalf("Task engine [%s]: unsupported desired state transition for resource [%s]: %s",
+			mtask.Arn, resource.GetName(), resource.StatusString(nextState))
+		return errors.New("impossible to transition to the resource state due to lack of transition function")
+	}
+	if err != nil {
+		seelog.Infof("Task engine [%s]: error transitioning resource [%s] to [%s]: %v",
+			mtask.Arn, resource.GetName(), resource.StatusString(nextState), err)
+	} else {
```

Comment:
just return`err` here, get rid of `else` and return `nil` at the end. Much easier to reason about and one less indentation level in the method.